### PR TITLE
Added support for Python FMCS functors

### PR DESCRIPTION
**Pull request by ptosco**
This PR adds support for user-defined Python FMCS functors and progress callbacks.
The pre-defined C++ compare functions can be called from Python too, so users do not have to reimplement Python versions of the already available C++ functionality (e.g., `CheckAtomRingMatch`, `CheckAtomCharge`, `CheckAtomChirality`, `CheckBondStereo`, `CheckBondRingMatch`), and can focus only on the custom behavior which is not available OOTB in the C++ layer.
I struggled to maintain backwards compatibility with the previous FMCS C++ implementation, so users who have written custom C++ `compare` functions won't need to change anything.
I had to add a `bool MatchIsotope` variable to `MCSAtomCompareParameters` or isotopic SMARTS could not have been written from Python unless the built-in `MCSAtomCompareIsotopes` is used.
From Python users do not have to use the horrible opaque `userData` C++ pointer to pass their own data to their `compare` functions. Instead, they may subclass a base class and reimplement the `compare()` method. This provides a cleaner interface and users can store data as well as add their own methods to the subclass.
I have added a fairly comprehensive Python test suite, which re-runs existing tests using Python implementations of the C++ `compare` functions and adds some specific tests.

### CODE REVIEW

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 1]
```diff
@@ -11,41 +11,383 @@
...
+    }
+  }
+  python::object getMCSProgressCallback() {
+    if (!pcud->pyMCSProgress.is_none()) {
+      return pcud->pyMCSProgress;
+    }
+    return python::object();
+  }
+  void clearRingTableCache() {
+    if (cfud->ringMatchTablesMols)
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: we're putting `{}` around single-line `if` statements now

File: Code/GraphMol/FMCS/FMCS.cpp [hunk 2]
```diff
@@ -24,6 +24,70 @@
...
+      AtomTyper = MCSAtomCompareElements;
+      break;
+    case AtomCompareIsotopes:
+      AtomTyper = MCSAtomCompareIsotopes;
+      break;
+    case AtomCompareAnyHeavyAtom:
+      AtomTyper = MCSAtomCompareAnyHeavyAtom;
+      break;
+    default:
+      throw std::runtime_error("Unknown AtomComparator");
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/FMCS.cpp]: It seems more explicit to do this with a `ValueErrorException`

File: Code/GraphMol/FMCS/FMCS.cpp [hunk 3]
```diff
@@ -24,6 +24,70 @@
...
+      BondTyper = MCSBondCompareAny;
+      break;
+    case BondCompareOrder:
+      BondTyper = MCSBondCompareOrder;
+      break;
+    case BondCompareOrderExact:
+      BondTyper = MCSBondCompareOrderExact;
+      break;
+    default:
+      throw std::runtime_error("Unknown BondComparator");
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/FMCS.cpp]: Same comment here about the `ValueErrorException`

File: Code/GraphMol/FMCS/FMCS.cpp [hunk 4]
```diff
@@ -24,6 +24,70 @@
...
+      break;
+    case AtomCompareAnyHeavyAtom:
+      AtomTyper = MCSAtomCompareAnyHeavyAtom;
+      break;
+    default:
+      throw std::runtime_error("Unknown AtomComparator");
+  }
+}
+
+void MCSParameters::setMCSAtomTyperFromConstChar(const char *atomComp) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/FMCS.cpp]: need a:
```
PRECONDITION(atomComp,"bad pointer");
```

File: Code/GraphMol/FMCS/FMCS.cpp [hunk 5]
```diff
@@ -24,6 +24,70 @@
...
+      break;
+    case BondCompareOrderExact:
+      BondTyper = MCSBondCompareOrderExact;
+      break;
+    default:
+      throw std::runtime_error("Unknown BondComparator");
+  }
+}
+
+void MCSParameters::setMCSBondTyperFromConstChar(const char *bondComp) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/FMCS.cpp]: same comment about the `PRECONDITION`

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 6]
```diff
@@ -11,41 +11,383 @@
...
+  void setBondCompareParameters(const MCSBondCompareParameters &value) {
+    p->BondCompareParameters = value;
+  }
+  std::string getInitialSeed() const {
+    return p->InitialSeed;
+  }
+  void setInitialSeed(const std::string &value) {
+    p->InitialSeed = value;
+  }
+  void setMCSAtomTyper(PyObject *atomComp) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: It's maybe not technically necessary here, but I always feel safer if a function taking a pointer that is going to be dereferenced has a `PRECONDITION()` on that pointer.

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 7]
```diff
@@ -11,41 +11,383 @@
...
+      res = python::object(atomTyperToComp.at(p->AtomTyper));
+    }
+    catch(const std::out_of_range&) {
+      PyErr_SetString(PyExc_TypeError,
+        "Unknown AtomTyper");
+      python::throw_error_already_set();
+    }
+    return res;
+  }
+  void setMCSBondTyper(PyObject *bondComp) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: Same comment about the PRECONDITION

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 8]
```diff
@@ -11,41 +11,383 @@
...
+      res = python::object(bondTyperToComp.at(p->BondTyper));
+    }
+    catch(const std::out_of_range&) {
+      PyErr_SetString(PyExc_TypeError,
+        "Unknown BondTyper");
+      python::throw_error_already_set();
+    }
+    return res;
+  }
+  void setMCSProgressCallback(PyObject *progress) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: Same comment about the PRECONDITION

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 9]
```diff
@@ -11,41 +11,383 @@
...
+    if (cfud->ringMatchTablesMols)
+      cfud->ringMatchTablesMols->clear();
+    if (cfud->ringMatchTables)
+      cfud->ringMatchTables->clear();
+  }
+private:
+  static bool MCSAtomComparePyFunc(const MCSAtomCompareParameters& p,
+                                   const ROMol& mol1, unsigned int atom1,
+                                   const ROMol& mol2, unsigned int atom2,
+                                   void* userData) {
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: Please add a `PRECONDITION` making sure that `userData` is not null.

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 10]
```diff
@@ -11,41 +11,383 @@
...
+    bool res = false;
+    {
+      PyGILStateHolder h;
+      res = python::call_method<bool>(cfud->pyAtomComp.ptr(), COMPARE_FUNC_NAME,
+                                      boost::ref(p), boost::ref(mol1),
+                                      atom1, boost::ref(mol2), atom2);
+    }
+    return res;
+  }
+  static bool MCSBondComparePyFunc(const MCSBondCompareParameters& p,
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: Please add a `PRECONDITION` making sure that `userData` is not null.

File: Code/GraphMol/FMCS/Wrap/rdFMCS.cpp [hunk 11]
```diff
@@ -11,41 +11,383 @@
...
+    bool res = false;
+    {
+      PyGILStateHolder h;
+      res = python::call_method<bool>(cfud->pyBondComp.ptr(), COMPARE_FUNC_NAME,
+                                      boost::ref(p), boost::ref(mol1),
+                                      bond1, boost::ref(mol2), bond2);
+    }
+    return res;
+  }
+  static bool MCSProgressCallbackPyFunc(const MCSProgressData& stat,
```
Comment (greglandrum) [inline Code/GraphMol/FMCS/Wrap/rdFMCS.cpp]: Please add a `PRECONDITION` making sure that `userData` is not null.

**Comment by ptosco** [comment]
@greglandrum Hi Greg, thanks, this all makes sense. I added a couple more missing `PRECONDITIONS` that I found in passing.

**Comment by greglandrum** [comment]
Thanks @ptosco !

**Pull request merged**